### PR TITLE
feat(tianmu) Supports conditional deletion of single table and multiple tables (#345)(#348)

### PR DESCRIPTION
**Pull request by konghaiya**
- #345 
- #348
- #451 

## Summary about this PR

Issue Number: close  (#345)(#348)

## Tests Check List

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [x] New Feature
- [ ] Bug Fix
- [x] Improvement
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

File: storage/tianmu/common/defs.h [hunk 1]
```diff
@@ -42,7 +42,7 @@ constexpr const char *COL_FILTER_HIST_DIR = "hist";
 constexpr const char *COL_KN_FILE = "KN";
 constexpr const char *COL_META_FILE = "META";
 constexpr const char *COL_DN_FILE = "DN";
-constexpr size_t COL_DN_FILE_SIZE = 10 * 1024 * 1024;  // given the pack size is 64K, we support up to 8G rows
+constexpr size_t COL_DN_FILE_SIZE = 11 * 1024 * 1024;  // given the pack size is 64K, we support up to 8G rows
```
Comment (RingsC) [inline storage/tianmu/common/defs.h]: Comment needed to explain the magic number.

File: storage/tianmu/core/column_share.cpp [hunk 2]
```diff
@@ -28,7 +28,7 @@
 namespace Tianmu {
 namespace core {
 // make sure the struct is not modified by mistake
-static_assert(sizeof(DPN) == 80, "Bad struct size of DPN");
+static_assert(sizeof(DPN) == 88, "Bad struct size of DPN");
```
Comment (RingsC) [inline storage/tianmu/core/column_share.cpp]: same as before, do not use magic num. if use, pls comments its meaning or explainations.

File: storage/tianmu/core/column_share.cpp [hunk 3]
```diff
@@ -191,8 +191,8 @@ void ColumnShare::init_dpn(DPN &dpn, const common::TX_ID xid, const DPN *from) {
 
 // should get page size at run time with sysconf(_SC_PAGE_SIZE) but for
 // efficiency...
-static constexpr size_t PAGE_SIZE = 4096;
-static constexpr size_t PAGE_CNT = 5;
+static constexpr size_t PAGE_SIZE = 2816;
```
Comment (RingsC) [inline storage/tianmu/core/column_share.cpp]: same as before

File: storage/tianmu/core/column_share.h [hunk 4]
```diff
@@ -43,9 +43,10 @@ struct COL_META {
 };
 
 struct alignas(128) COL_VER_HDR_V3 {
-  uint64_t nr;  // no. of records
-  uint64_t nn;  // no. of nulls
-  uint64_t np;  // no. of packs
+  uint64_t nr;       // no. of records
```
Comment (RingsC) [inline storage/tianmu/core/column_share.h]: Try to use a better naming style. This is a bad example.

File: storage/tianmu/core/compilation_tools.cpp [hunk 5]
```diff
@@ -116,8 +116,8 @@ int OperationUnmysterify(Item *item, common::ColOperation &oper, bool &distinct,
         case Item_sum::GROUP_CONCAT_FUNC:
           distinct = ((Item_func_group_concat *)item)->get_distinct();
           TIANMU_LOG(LogCtl_Level::DEBUG, "group_concat distinct %d, sepertator %s, direction %d", distinct,
-                      ((Item_func_group_concat *)item)->get_separator()->c_ptr(),
-                      ((Item_func_group_concat *)item)->direction());
+                     ((Item_func_group_concat *)item)->get_separator()->c_ptr(),
```
Comment (RingsC) [inline storage/tianmu/core/compilation_tools.cpp]: Do NOT USE c style, c++ type casting style.

File: storage/tianmu/core/dpn.h [hunk 6]
```diff
@@ -47,6 +48,7 @@ struct DPN final {
   uint64_t len;        // data length
   uint32_t nr;         // number of records
```
Comment (RingsC) [inline storage/tianmu/core/dpn.h]: bad naming style

File: storage/tianmu/core/dpn.h [hunk 7]
```diff
@@ -39,6 +39,7 @@ struct DPN final {
   uint8_t synced : 1;  // if the pack data in memory is up to date with the
                        // version on disk
   uint8_t null_compressed : 1;
+  uint8_t delete_compressed : 1;
   uint8_t data_compressed : 1;
```
Comment (RingsC) [inline storage/tianmu/core/dpn.h:43]: Notice: The cache line alignment

File: storage/tianmu/core/pack.cpp [hunk 8]
```diff
@@ -21,11 +21,19 @@
...
+#include "core/column_share.h"
+#include "core/value.h"
+#include "loader/value_cache.h"
+#include "system/tianmu_file.h"
 namespace Tianmu {
 namespace core {
 Pack::Pack(DPN *dpn, PackCoordinate pc, ColumnShare *s) : s(s), dpn(dpn) {
   NULLS_SIZE = (1 << s->pss) / 8;
   nulls = std::make_unique<uint32_t[]>(NULLS_SIZE / sizeof(uint32_t));
+  deletes = std::make_unique<uint32_t[]>(NULLS_SIZE / sizeof(uint32_t));
```
Comment (RingsC) [inline storage/tianmu/core/pack.cpp]: Why is NULL_SIZE ?

File: storage/tianmu/core/group_table.cpp [hunk 9]
```diff
@@ -370,10 +370,11 @@ void GroupTable::Initialize(int64_t max_no_groups, bool parallel_allowed) {
   Transaction *m_conn = current_txn_;
   double size_mb = (double)vm_tab->ByteSize();
   size_mb =
-      (size_mb > 1024 ? (size_mb > 1024 * 1024 ? int64_t(size_mb / 1024 * 1024) : int64_t(size_mb / 1024) / 1024.0) : 0.001);
-  rc_control_.lock(m_conn->GetThreadID()) << "GroupTable begin, initialized for up to " << max_no_groups << " groups, "
-                                        << grouping_buf_width << "+" << total_width - grouping_buf_width << " bytes ("
-                                        << size_mb << " MB)" << system::unlock;
+      (size_mb > 1024 ? (size_mb > 1024 * 1024 ? int64_t(size_mb / 1024 * 1024) : int64_t(size_mb / 1024) / 1024.0)
+                      : 0.001);
```
Comment (RingsC) [inline storage/tianmu/core/group_table.cpp]: 0.001 means what?

**Review by RingsC — CHANGES_REQUESTED** [review]

File: storage/tianmu/core/pack.cpp [hunk 10]
```diff
@@ -65,5 +75,34 @@ void Pack::Release() {
 bool Pack::ShouldNotCompress() const {
   return (dpn->nr < (1U << s->pss)) || (s->ColType().GetFmt() == common::PackFmt::NOCOMPRESS);
 }
+
+bool Pack::CompressedNullsOrDeletes(mm::MMGuard<uchar> &comp_buf, uint &comp_buf_size,
+                                    std::unique_ptr<uint32_t[]> &ptr_buf, uint32_t &dpn_num1) {
+  comp_buf_size = ((dpn->nr + 7) / 8);
```
Comment (RingsC) [inline storage/tianmu/core/pack.cpp]: The magic number?

File: storage/tianmu/core/pack.cpp [hunk 11]
```diff
@@ -65,5 +75,34 @@ void Pack::Release() {
...
+
+  comp_buf =
+      mm::MMGuard<uchar>((uchar *)alloc((comp_buf_size + 2) * sizeof(char), mm::BLOCK_TYPE::BLOCK_TEMPORARY), *this);
```
Comment (RingsC) [inline storage/tianmu/core/pack.cpp:89]: how about: 1) comp_buf_size is too large that make runing out of the memory.

File: storage/tianmu/core/pack.h [hunk 12]
```diff
@@ -50,6 +51,7 @@ class Pack : public mm::TraceableObject {
   virtual void LoadDataFromFile(system::Stream *fcurfile) = 0;
   virtual void Save() = 0;
   virtual void UpdateValue(size_t i, const Value &v) = 0;
+  virtual void DeleteByRow(size_t i) = 0;
```
Comment (RingsC) [inline storage/tianmu/core/pack.h:54]: By row, but input param is "size_t i" ? How to do delete with an iteger value.

File: storage/tianmu/core/pack.h [hunk 13]
```diff
@@ -71,6 +73,24 @@ class Pack : public mm::TraceableObject {
     if (dpn->nn == dpn->nr) return true;
     return ((nulls[i >> 5] & ((uint32_t)(1) << (i % 32))) != 0);
   }
+
+  void SetDelete(int i) {
+    int mask = 1 << (i % 32);
+    ASSERT((deletes[i >> 5] & mask) == 0);
```
Comment (RingsC) [inline storage/tianmu/core/pack.h]: meaning of 32, and 5 ?

Comment (RingsC) [reply storage/tianmu/core/pack.h]: `i` is what ?

File: storage/tianmu/core/pack.h [hunk 14]
```diff
@@ -71,6 +73,24 @@ class Pack : public mm::TraceableObject {
...
+  }
+
+  void UnsetDelete(int i) {
+    int mask = ~(1 << (i % 32));
+    ASSERT(IsDelete(i), "already delete!");
+    deletes[i >> 5] &= mask;
+  }
+
+  bool IsDelete(int i) const {
+    if (dpn->ndelete == dpn->nr) return true;
```
Comment (RingsC) [inline storage/tianmu/core/pack.h]: IsDelete? means what? have done  or not yet.

**Review by RingsC — CHANGES_REQUESTED** [review]

File: storage/tianmu/index/rc_table_index.h [hunk 15]
```diff
@@ -56,6 +56,7 @@ class RCTableIndex final {
   common::ErrorCode RenameIndexTable(const std::string &from, const std::string &to);
   common::ErrorCode InsertIndex(core::Transaction *tx, std::vector<std::string_view> &fields, uint64_t row);
   common::ErrorCode UpdateIndex(core::Transaction *tx, std::string_view &nkey, std::string_view &okey, uint64_t row);
+  common::ErrorCode DeleteIndex(core::Transaction *tx, std::string_view &okey, uint64_t row);
```
Comment (RingsC) [inline storage/tianmu/index/rc_table_index.h]: we dont understand the literal meaning of `okey` from the variable name.

File: storage/tianmu/index/rc_table_index.cpp [hunk 16]
```diff
@@ -206,6 +206,24 @@ common::ErrorCode RCTableIndex::UpdateIndex(core::Transaction *tx, std::string_v
...
+common::ErrorCode RCTableIndex::DeleteIndex(core::Transaction *tx, std::string_view &okey,
+                                            uint64_t row) {
+  StringWriter value, packkey;
+  std::vector<std::string_view> ofields;
+
+  ofields.emplace_back(okey);
+
+  rocksdb_key_->pack_key(packkey, ofields, value);
+  common::ErrorCode rc = CheckUniqueness(tx, {(const char *)packkey.ptr(), packkey.length()});
+  if (rc == common::ErrorCode::DUPP_KEY) {
```
Comment (RingsC) [inline storage/tianmu/index/rc_table_index.cpp]: the resutl of `DeleteIndex` is return value of `CheckUniqueness`, not the return value of deletion.  if the rc is `SUCCESS`, what will happens?

Comment (RingsC) [reply storage/tianmu/index/rc_table_index.cpp]: Thinking these logicall flow carefully.

Comment (RingsC) [reply storage/tianmu/index/rc_table_index.cpp]: How about the un-unique key?

Comment (RingsC) [reply storage/tianmu/index/rc_table_index.cpp]: No c-style type casting.

File: storage/tianmu/core/rc_attr.cpp [hunk 17]
```diff
@@ -976,8 +980,68 @@ void RCAttr::UpdateData(uint64_t row, Value &v) {
...
+  }
+  return false;
+}
+
+void RCAttr::DeleteData(uint64_t row) {
+  auto pn = row2pack(row);
+  FunctionExecutor fe([this, pn]() { LockPackForUse(pn); }, [this, pn]() { UnlockPackFromUse(pn); });
+
+  // primary key process
+  DeleteIfIndex(row, ColId());
```
Comment (RingsC) [inline storage/tianmu/core/rc_attr.cpp]: How do we know the satus of these operations, `DeleteIfIndex` , `CopyPackForWrite`, etc.

File: storage/tianmu/core/column_share.cpp [hunk 18]
```diff
@@ -27,8 +27,24 @@
 
 namespace Tianmu {
 namespace core {
+
+//Current DNP size
```
Comment (RingsC) [inline storage/tianmu/core/column_share.cpp]: Why is 88, rather than any arbitrary number.

Comment (RingsC) [reply storage/tianmu/core/column_share.cpp]: I think that we should write some words to explain all of these maigc numbers.

File: storage/tianmu/core/column_share.cpp [hunk 19]
```diff
@@ -168,7 +184,7 @@ void ColumnShare::init_dpn(DPN &dpn, const common::TX_ID xid, const DPN *from) {
     dpn = *from;
   } else {
     dpn.reset();
```
Comment (RingsC) [inline storage/tianmu/core/column_share.cpp:192]: I think that the following codes can refactor.  
```
reset ()
{
     ...
    dpn.dataAddress = DPN_INVALID_ADDR;
    if (pt == common::PackType::INT) {
      dpn.min_i = common::PLUS_INF_64;
      dpn.max_i = common::MINUS_INF_64;
     ...
}
```

File: storage/tianmu/core/column_share.cpp [hunk 20]
```diff
@@ -189,15 +205,6 @@ void ColumnShare::init_dpn(DPN &dpn, const common::TX_ID xid, const DPN *from) {
...
-static constexpr size_t PAGE_SIZE = 4096;
-static constexpr size_t PAGE_CNT = 5;
-static constexpr size_t ALLOC_UNIT = PAGE_CNT * PAGE_SIZE;
-
-static_assert(ALLOC_UNIT % sizeof(DPN) == 0);
-
-static constexpr size_t DPN_INC_CNT = ALLOC_UNIT / sizeof(DPN);
 
 int ColumnShare::alloc_dpn(common::TX_ID xid, const DPN *from) {
   for (uint32_t i = 0; i < cap; i++) {
```
Comment (RingsC) [inline storage/tianmu/core/column_share.cpp]: `capacity`  is better than `cap`.

File: storage/tianmu/core/rc_attr.cpp [hunk 21]
```diff
@@ -1251,5 +1315,25 @@ void RCAttr::UpdateIfIndex(uint64_t row, uint64_t col, const Value &v) {
     }
   }
 }
+
+void RCAttr::DeleteIfIndex(uint64_t row, uint64_t col) {
```
Comment (RingsC) [inline storage/tianmu/core/rc_attr.cpp]: can not descibe the function totally.    `DeleteByPrimaryKey` ?

File: storage/tianmu/core/rc_attr.cpp [hunk 22]
```diff
@@ -1251,5 +1315,25 @@ void RCAttr::UpdateIfIndex(uint64_t row, uint64_t col, const Value &v) {
...
+  std::shared_ptr<index::RCTableIndex> tab = ha_rcengine_->GetTableIndex(path);
+  // col is not primary key
+  if (!tab) return;
+  std::vector<uint> keycols = tab->KeyCols();
+  if (std::find(keycols.begin(), keycols.end(), col) == keycols.end()) return;
+
+  if (GetPackType() == common::PackType::STR) {
+    auto vold = GetValueString(row);
+    std::string_view okey(vold.val, vold.size());
+    tab->DeleteIndex(current_txn_, okey, row);
```
Comment (RingsC) [inline storage/tianmu/core/rc_attr.cpp]: these codes can be refined, redundent lines removed

File: storage/tianmu/core/rcattr_exqp.cpp [hunk 23]
```diff
@@ -287,24 +289,50 @@ common::ErrorCode RCAttr::EvaluateOnIndex_BetweenString_UTF(MIUpdatingIterator &
   return rv;
 }
 
+void RCAttr::EvaluatePack_IsNoDelete(MIUpdatingIterator &mit, int dim) {
+  MEASURE_FET("RCAttr::EvaluatePack_IsNoDelete(...)");
+  int64 curRow = mit[dim];
+  auto pack = row2pack(curRow);
+  auto const &dpn(get_dpn(pack));
+  if (dpn.numOfDeleted > 0) {
```
Comment (RingsC) [inline storage/tianmu/core/rcattr_exqp.cpp:296]: Refine these lines. and make it more concise.

File: storage/tianmu/handler/tianmu_handler.cpp [hunk 24]
```diff
@@ -534,9 +534,35 @@ int TianmuHandler::update_row(const uchar *old_data, uchar *new_data) {
  */
 int TianmuHandler::delete_row([[maybe_unused]] const uchar *buf) {
   DBUG_ENTER(__PRETTY_FUNCTION__);
-  DBUG_RETURN(HA_ERR_WRONG_COMMAND);
-}
+  int ret = HA_ERR_INTERNAL_ERROR;
+  auto org_bitmap = dbug_tmp_use_all_columns(table, table->write_set);
```
Comment (RingsC) [inline storage/tianmu/handler/tianmu_handler.cpp:538]: be aware of multi-version issues.

**Review by RingsC — APPROVED** [review]

File: storage/tianmu/core/rc_attr.cpp (hunk 17 above)
Comment (konghaiya) [reply storage/tianmu/core/rc_attr.cpp]: Although these two functions have no return value, exception capture is carried out in the logic of the function. If an exception occurs, we can judge the execution status of the function through exception capture.

File: storage/tianmu/handler/tianmu_handler.cpp (hunk 24 above)
Comment (konghaiya) [reply storage/tianmu/handler/tianmu_handler.cpp:538]: This is indeed a problem worth considering. 
Our Tianmu engine currently supports multiple versions when update and delete

**Review by RingsC — APPROVED** [review]
LGTM

**Review by adofsauron — APPROVED** [review]
LGTM

**Pull request merged**